README.md
[![Build Status](https://travis-ci.org/smelukov/NanoEqual.svg?branch=master)](https://travis-ci.org/smelukov/NanoEqual)
[![npm version](https://badge.fury.io/js/nano-equal.svg)](https://badge.fury.io/js/nano-equal)

## NanoEqual
Ultra fast and compact implementation of deep equal without any production dependencies.

### Usage
```shell
npm install nano-equal --save
```

```javascript
var nanoEqual = require('nano-equal');

if (nanoEqual(a, b)) {
    //....
}
```

### What is that?
This is a compact and fast implementation of deep equal.

Deep equal is an algorithm that comparing two values. If the values are a scalar (string, bool, number), then comparing will be performed thru === operator.

If the values are an object (object, array, function), then comparing will be performed recursively.

Following of the object properties is not important:
```javascript
var a = {prop1: 'some', prop2: 'some'};
var b = {prop2: 'some', prop1: 'some'};

nanoEqual(a, b); // true
```

But following of the array elements is important:
```javascript
var a = [1, 2, 3];
var b = [3, 2, 1];

nanoEqual(a, b); // false
```

> NaN values and some types of recursion are supported.

### Benchmark

The list below is showing the performance comparison nano-equal with other libs:

```
nanoEqual: 1362.813ms
underscore: 3791.308ms
lodash: 7830.107ms
nodejs: 8272.956ms
```


CHANGELOG.md
## 2.0.1 (August 24, 2017)

- fixed array-like comparison

## 2.0.0 (August 24, 2017)

- performance improvements
- array recursion check
- removed gulp and webpack

## 1.0.2 (March 10, 2016)

- fix recursion object check

# 1.0.1 (March 10, 2016)

- micro optimization

# 1.0.0 (March 10, 2016)

- initial release


test/main.js
var assert = require('chai').assert;
var nanoEqual = require('../src/index');

var objA = {
    nan: NaN,
    prop1: 'value1',
    prop2: 'value2',
    prop3: 'value3',
    prop4: {
        subProp1: 'sub value1',
        subProp2: {
            subSubProp1: 'sub sub value1',
            subSubProp2: [1, 2, { prop2: 1, prop: 2 }, 4, 5]
        }
    },
    prop5: 1000,
    prop6: new Date(2016, 2, 10)
};

var objB = {
    nan: NaN,
    prop5: 1000,
    prop3: 'value3',
    prop1: 'value1',
    prop2: 'value2',
    prop6: new Date('2016/03/10'),
    prop4: {
        subProp2: {
            subSubProp1: 'sub sub value1',
            subSubProp2: [1, 2, { prop2: 1, prop: 2 }, 4, 5]
        },
        subProp1: 'sub value1'
    }
};

var objBWithAddProp = {
    prop5: 1000,
    prop3: 'value3',
    prop1: 'value1',
    prop2: 'value2',
    prop6: new Date('2016/03/10'),
    prop4: {
        subProp2: {
            subSubProp1: 'sub sub value1',
            subSubProp2: [1, 2, { prop2: 1, prop: 2 }, 4, 5]
        },
        subProp1: 'sub value1'
    },
    addProp: ';('
};

var recA = {};
var recB = {};

recA.a = recA;
recB.a = recB;

var recA2 = {};
var recB2 = {};

recA2.a = recA2;
recB2.a = recA2;

var recA3 = {};
var recB3 = {};

recA3.a = recB3;
recB3.a = recA3;

describe('Functional', function() {
    it('should working correctly', function() {
        assert.isTrue(nanoEqual(objA, objA));
        assert.isTrue(nanoEqual(objA, objB));
        assert.isFalse(nanoEqual(objA, objBWithAddProp));

        assert.isFalse(nanoEqual(recA, recB));
        assert.isTrue(nanoEqual(recA2, recB2));
        assert.isFalse(nanoEqual(recA3, recB3));
    });
});


gulpfile.js
var concat = require('gulp-concat'),
    del = require('del'),
    eslint = require('gulp-eslint'),
    gulp = require('gulp'),
    gutil = require('gulp-util'),
    header = require('gulp-header'),
    jsdoc2md = require('gulp-jsdoc-to-markdown'),
    jscs = require('gulp-jscs'),
    mocha = require('gulp-mocha'),
    webpack = require('webpack-stream');

gulp.task('default', ['clean', 'eslint', 'jscs', 'build', 'license', 'tests', 'docs']);

gulp.task('clean', function() {
    del.sync('docs', {force: true});

    return del.sync('build', {force: true});
});

gulp.task('jscs', function() {
    return gulp.src(['src/**/*.js', 'test/**/*.js'])
        .pipe(jscs())
        .pipe(jscs.reporter())
        .pipe(jscs.reporter('fail'));
});

gulp.task('eslint', function() {
    return gulp.src('./src/**/*.js')
        .pipe(eslint())
        .pipe(eslint.format())
        .pipe(eslint.failAfterError());
});

gulp.task('tests', ['build'], function() {
    return gulp.src('test/**/*.js')
        .pipe(mocha());
});

gulp.task('build', ['clean'], function() {
    return webpack(require('./webpack.config.js'))
        .pipe(gulp.dest('./build/'));
});

gulp.task('license', ['build'], function() {
    var banner = ['/**',
        ' * <%= pkg.name %> - <%= pkg.description %>',
        ' * @version v<%= pkg.version %>',
        ' * @link <%= pkg.homepage %>',
        ' * @license <%= pkg.license %>',
        ' */',
        ''].join('\n');

    return gulp.src('./build/nanoEqual.min.js')
        .pipe(header(banner, {pkg: require('./package.json')}))
        .pipe(gulp.dest('./build/'));
});

gulp.task('docs', ['build'], function() {
    return gulp.src('src/*.js')
        .pipe(concat('README.md'))
        .pipe(jsdoc2md())
        .on('error', function(err) {
            gutil.log('jsdoc2md failed:', err.message);
        })
        .pipe(gulp.dest('docs'));
});


src/index.js
/**
 * @licence
 * @author Sergey Melyukov @smelukov
 */

/**
 * Is value like array?
 *
 * @param {*} a
 * @returns {boolean}
 */
function isArrayLike(a) {
    if (Array.isArray(a)) {
        return true;
    }

    var len = a.length;

    if (typeof len === 'number' && len > -1) {
        if (len) {
            return 0 in a && len - 1 in a;
        }

        return true;
    }

    return false;
}

/**
 * Get type of value.
 *
 * @param {*} a
 * @returns {string}
 */
function getType(a) {
    var type = typeof a;

    if (type === 'object') {
        if (a === null) {
            return 'null';
        } else if (isArrayLike(a)) {
            return 'array';
        } else if (a.constructor === Object) {
            return 'pure-object';
        }

        return 'object';
    }

    return type;
}

/**
 * Deep equal of the values
 *
 * @param {*} a
 * @param {*} b
 * @returns {boolean}
 */
function nanoEqual(a, b) {
    if (a === b) {
        return true;
    }

    // is nan
    if (a !== a && b !== b) { // eslint-disable-line no-self-compare
        return true;
    }

    var typeA = getType(a);
    var typeB = getType(b);

    if (typeA !== typeB) {
        return false;
    }

    if (typeA === 'pure-object') {
        if (a === b) {
            return true;
        }

        var keys = Object.keys(a);

        for (var i = 0, l = keys.length; i < l; i++) {
            var key = keys[i];

            if (!b.hasOwnProperty(keys[i])) {
                return false;
            }

            var valA = a[key];
            var valB = b[key];

            // handle recursion
            if (valA === a || valB === b || valA === b || valB === a) {
                return valA === valB;
            }

            if (!nanoEqual(valA, valB)) {
                return false;
            }
        }

        return true;
    } else if (typeA === 'array') {
        if (a.length === b.length) {
            for (var j = 0; j < a.length; j++) {
                var elA = a[j];
                var elB = b[j];

                // handle recursion
                if (elA === a || elB === b || elA === b || elB === a) {
                    return elA === elB;
                }

                if (!nanoEqual(elA, elB)) {
                    return false;
                }
            }
        }

        return true;
    } else if (typeA === 'object') {
        if (a.valueOf !== Object.prototype.valueOf() && b.valueOf !== Object.prototype.valueOf()) {
            return a.valueOf() === b.valueOf();
        }

        if (a.toString !== Object.prototype.toString() && b.toString !== Object.prototype.toString()) {
            return a.toString() === b.toString();
        }
    }

    return false;
}

module.exports = nanoEqual;


example/comparison.js
var nanoEqual = require('./../src');
var lodashEqual = require('lodash').isEqual;
var underscoreEqual = require('underscore').isEqual;
var assert = require('assert');
var nodeEqual = function(a, b) {
    try {
        assert.deepEqual(a, b);
        return true;
    } catch (e) {
        return false;
    }
};
var objA = {
    prop1: 'value1',
    prop2: 'value2',
    prop3: 'value3',
    prop4: {
        subProp1: 'sub value1',
        subProp2: {
            subSubProp1: 'sub sub value1',
            subSubProp2: [1, 2, { prop2: 1, prop: 2 }, 4, 5]
        }
    },
    prop5: 1000,
    prop6: new Date(2016, 2, 10)
};

var objB = {
    prop5: 1000,
    prop3: 'value3',
    prop1: 'value1',
    prop2: 'value2',
    prop6: new Date('2016/03/10'),
    prop4: {
        subProp2: {
            subSubProp1: 'sub sub value1',
            subSubProp2: [1, 2, { prop2: 1, prop: 2 }, 4, 5]
        },
        subProp1: 'sub value1'
    }
};

function testEqual(name, fn, a, b) {
    console.time(name);

    for (var i = 0; i < 1000000; i++) {
        if (!fn(a, b)) {
            console.timeEnd(name);
            console.log('not equal');
            return;
        }
    }

    console.timeEnd(name);
}

testEqual('nanoEqual', nanoEqual, objA, objB);
testEqual('nodejs', nodeEqual, objA, objB);
testEqual('lodash', lodashEqual, objA, objB);
testEqual('underscore', underscoreEqual, objA, objB);
